Add support for config file, using Viper

danielhunt opened the pull request:
Hey @zramsay - I randomly came across your repo while starting out learning Go, and loved that you made this application available for me to look through and test myself.

Skimming through your issues list, I figured it would be helpful if I added config-file support, so this fixes #2 

If the code quality here isn't up to scratch, let me know and I'll improve what I can ... but go easy. I'm new to Go ;)

danielhunt committed:
Add support for config file, using Viper

zramsay commented:
@danielhunt, this is nice work, thanks. it's *mostly* good to go, although there are handful of fixes that i'll comment on directly in the code. As well, the primary intention for issue #2 was to handle [this implementation specific part of the code](https://github.com/CoinCulture/point-of-sale/blob/master/main.go#L420-L427), so we'd want to get that added and the logic cleaned up.

zramsay commented on config.go:
```diff
@@ -0,0 +1,62 @@
+
+func (c DbConfig) getAddr() string {
+	if c.net == "tcp" {
+		return c.host + ":" + strconv.Itoa(c.port)
```
concatenation should use `fmt.Sprintf`, for example:
```
return fmt.Sprintf("%s:%s", c.host, strconv.Itoa(c.port)
```

zramsay commented on config.go:
```diff
@@ -0,0 +1,62 @@
+
+	err := viper.ReadInConfig()
+	if err != nil {
+		fmt.Println("Config file (./config/app.[json|yaml|toml]) not found - abort")
```
let's call the file `config.toml`, have it in the main repo directory, and restrict that it be `.toml` only with a check in this function

danielhunt replied:
I don't see any reason to enforce `toml`, but hey, it's your project ;)

zramsay replied:
mostly for simplicity and to reduce the number of concepts a beginner needs to grasp

danielhunt replied:
Ya - no worries. I've changed the messaging to reflect this

zramsay commented on config.go:
```diff
@@ -0,0 +1,62 @@
+		os.Exit(1)
+	}
+
+	viper.SetDefault("env."+*env+".http.port", 8080)
```
fmt.Sprintf

zramsay commented on config/app.yaml:
```diff
@@ -0,0 +1,13 @@
+#      port : 3306
+      name : "myBusiness"
+      user : "root"
+      pass : "rootpassword"
```
we should force users to enter password via the command line when running the app, rather than store in plaintext. this can be left as an option here, but throw a warning/prompt that requires over-riding. w see also #4, which is intended to add a hidden password prompt when running the app

danielhunt replied:
I think I'll remove this line entirely and we can look at enforcement inline in the app itself

zramsay commented on main.go:
```diff
@@ -655,13 +655,17 @@ func insertNewItems(w http.ResponseWriter, r *http.Request) {
 
 func main() {
-	password := flag.String("password", "", "password for mysql db")
+	env := flag.String("env", "development", "Environment [development|staging|production]")
```
let's keep the password flag, although i think i sort of see where you're going with this flag (which we can also keep if you give it more reason for being

danielhunt replied:
I was actually toying with the idea of enforcing password supply via an ENV variable:

```
DB_PASS=supersecret ./point-of-sale
```

That would do away with the concern about password storage in the app itself, as well as allowing for scripted application starts.

As for allowing an `env` flag to toggle a few settings at a time, I do think that the ability to do so is a good idea, but as elsewhere, this is your baby.
I'll rip it out if you think it's just adding confusion.

danielhunt replied:
Also, *not* having a password set is a valid option - can't forget that ;)

zramsay commented on main.go:
```diff
@@ -655,13 +655,17 @@ func insertNewItems(w http.ResponseWriter, r *http.Request) {
-		DBName: "myBusiness",
-		Passwd: *password,
+		User:   config.db.user,
+		DBName: config.db.databaseName,
```
awesome

zramsay commented on main.go:
```diff
@@ -706,8 +710,9 @@ func main() {
-	err = http.ListenAndServe(":8081", nil)
+	httpPortStr := strconv.Itoa(config.http.port)
+	fmt.Printf("POS app started, listening on port %v\n", httpPortStr)
+	err = http.ListenAndServe(":"+httpPortStr, nil)
```
sweet :) ... this also needs fmt.Sprintf

zramsay commented:
this would need some tests too, see for example https://github.com/CoinCulture/claw/pull/11/files#diff-e769ee28e26ec40f5b7f9640d22a42f3

danielhunt committed:
Address PR review comments
- change all string concatenation to use sprintf
- remove concept of environments. the app just doesn't need it yet
- remove password from config file. that can be addressed as a different piece of work
- change config approach to expect a toml file

danielhunt committed:
assume defaults where possible

danielhunt commented:
@zramsay I've addressed all comments - tests haven't been added because that setup wasn't in place
Happy to add them here if you want me to.

Can you have another look and see if it suits now?

zramsay commented:
LGTM, though still missing the part for dealing with foods-to-not-print and foods-that-don't-buzz

danielhunt commented:
@zramsay I'd argue that the food print/buzz stuff shouldn't be part of this PR anyway
That work is an extension to an already-working config system, imo

zramsay commented:
edited base to be `develop` branch, to add some git flow
